pkg/nvpci/nvpci.go

```diff
@@ -90,7 +90,12 @@ type NvidiaPCIDevice struct {
 	NumaNode   int
 	Config     *ConfigSpace
 	Resources  MemoryResources
-	IsVF       bool
+
+	// SRIOV info
+	IsVF     bool
+	TotalVFs *uint64
+	NumVFs   *uint64
+	PhysFn   *NvidiaPCIDevice
```

Comment:
Does it make sense to put this into a separate struct and use a `nil` value of the struct to indicate that it is *not* a VF, i.e.:

```
type NvidiaPCIDevice struct {
	...
	SriovInfo *SriovInfo
}

type SriovInfo struct {
	PhysFn   *NvidiaPCIDevice
	TotalVFs uint64
	NumVFs   uint64
}

func (p *nvpci) IsVF() bool {
	return p.SriovInfo != nil
}
```

Reply:
So the problem is that `PhysFn` is only valid for VFs and `TotalVFs` and `NumVFs` only for PFs. We can do it however like that and remove `IsVF` boolean:
```go
type NvidiaPCIDevice struct {
	...
	SriovInfo *SriovInfo
	PhysFn   *NvidiaPCIDevice
}

type SriovInfo struct {
	TotalVFs uint64
	NumVFs   uint64
}

func (p *nvpci) IsVF() bool {
	return p.PhysFn != nil
}
```
dunno if it makes it more robust.

Reply:
I see. It should follow a one-of pattern then:
```
// Only one should be set at any given time
type SriovInfo struct {
  PhysicalFunction *SriovPhysicalFunction
  VirtualFunction *SriovVirtualFunction
}

type SriovPhysicalFunction struct {
  	TotalVFs uint64
	NumVFs   uint64
}

type SriovVirtualFunction struct {
    PhysicalFunction *NvidiaPCIDevice
}

type NvidiaPCIDevice struct {
	...
	SriovInfo *SriovInfo
}

func (p *nvpci) IsPF() bool {
	return p.PhysicalFunction != nil
}

func (p *nvpci) IsVF() bool {
	return p.VirtualFunction != nil
}
```

Reply:
Sounds good. Will make the change shortly.

Reply:
Changed.